### PR TITLE
Support slicing tree model

### LINKED ISSUE #4052: [feature proposal] Load best model after early stopping
Could you please implement a way to load the best model after ealy stopping? LightGBM is also doing this by default. See here: https://lightgbm.readthedocs.io/en/latest/Python-Intro.html#early-stopping

That would be a great improvement to my hyperparameter optimization workflow.

Thanks
Philip

### LINKED ISSUE #5531: Removing trees from a model
As a follow up to https://github.com/dmlc/xgboost/issues/2281, I think that it would be great if XGBoost supported the removal of trees from a model. 

It would allow the implementation (with callbacks) of adaptive learning rate strategies that need to abandon some trees before adjusting the learning rate. More generally, it would allow to implement various "ensemble pruning" methods.

### PR DESCRIPTION
**Pull request by trivialfis**
This PR is meant the end the confusion around `best_ntree_limit` and unify model slicing.  We have multi-class and random forests, asking users to understand how to set `ntree_limit` is difficult and error prone.

Close #5531 Close #4052.

* Implement the `save_best` option in early stopping.
* Negative index is not supported.

### CODE REVIEW

**Comment by trivialfis** [comment]
Currently I treat out of bound error specially raise `IndexError` but  `ValueError` for other issues.  Should I always raise `IndexError` instead?  Feel free to provide suggestions . ;-)

**Comment by trivialfis** [comment]
Pasting offline conversion with @hcho3 here.  The trees in xgboost can be considered as a 3 dim tensor.  First dim is the number of boosting round, second is number of classes, and the last is size of forest.  This PR supports only slicing the first dim (number of boosted rounds), it's possible to support slicing other dimensions. but to us that seems to be over engineering.

**Review by RAMitchell — APPROVED** [review]
This is a great new feature. Any plans to deprecate `ntree_limit` throughout the code base in favour of the new terminology?

**Comment by trivialfis** [comment]
There are other language bindings out there.  I need to go over them to deprecate the parameter.

File: include/xgboost/gbm.h [hunk 1]
```diff
@@ -60,6 +60,16 @@ class GradientBooster : public Model, public Configurable {
    * \param fo output stream
    */
   virtual void Save(dmlc::Stream* fo) const = 0;
+  /*!
+   * \brief Slice the model.
+   * \param layer_begin Begining of boosted tree layer used for prediction.
+   * \param layer_end   End of booster layer. 0 means do not limit trees.
+   * \param out         Output gradient booster
+   */
```
Comment (hcho3) [inline include/xgboost/gbm.h:68]:
```suggestion
  /*!
   * \brief Slice a model using boosting index. The slice m:n indicates taking all trees
 *        that were fit during the boosting rounds m, (m+1), (m+2), ..., (n-1).
   * \param layer_begin Begining of boosted tree layer used for prediction.
   * \param layer_end   End of booster layer. 0 means do not limit trees.
   * \param out         Output gradient booster
   */
```

File: tests/python/test_basic_models.py [hunk 2]
```diff
@@ -325,3 +325,69 @@ def validate_model(parameters):
...
+        self.assertRaises(IndexError, lambda: booster[:num_boost_round+1])
+        self.assertRaises(ValueError, lambda: booster[1, 2])  # too many dims
+
+        def assign():
+            booster[...:end] = booster
+        self.assertRaises(TypeError, assign)
+
+    def test_slice(self):
+        self.run_slice('gbtree')
+        self.run_slice('dart')
```
Comment (hcho3) [inline tests/python/test_basic_models.py:393]: Can we use `@pytest.mark.parameterize` instead?

```python
@pytest.mark.parameterize(booster, ['gbtree', 'dart'])
def run_slice(self, booster):
```

See examples at [Parameterizing tests](https://docs.pytest.org/en/6.1.2/example/parametrize.html)

File: src/gbm/gbtree.cc [hunk 3]
```diff
@@ -398,6 +398,38 @@ void GBTree::SaveModel(Json* p_out) const {
...
+  std::vector<int32_t> &out_trees_info = out_model.tree_info;
+  out_trees_info.resize(layer_trees * n_layers);
+  out_model.param.num_trees = out_model.trees.size();
+  CHECK(this->model_.trees_to_update.empty());
+
+  *out_of_bound = detail::SliceTrees(
+      layer_begin, layer_end, step, this->model_, tparam_, layer_trees,
+      [&](auto const &in_it, auto const &out_it) {
+        auto new_tree =
+            std::make_unique<RegTree>(*this->model_.trees.at(in_it));
```
Comment (hcho3) [inline src/gbm/gbtree.cc:426]: Do we have assurance that the implicitly generated copy constructor `RegTree(const RegTree&)` behaves correctly?

Comment (trivialfis) [reply src/gbm/gbtree.cc:426]: Added tests with prediction.

File: tests/python/test_basic_models.py (hunk 2 above)
Comment (trivialfis) [reply tests/python/test_basic_models.py:393]: It seems not being compatible with class  method:
```
TypeError: test_slice() missing 1 required positional argument: 'booster'
```

File: doc/python/model.rst [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+.. code-block:: python
+
+    from sklearn.datasets import make_classification
+    num_classes = 3
+    X, y = make_classification(n_samples=1000, n_informative=5,
+                               n_classes=num_classes)
+    dtrain = xgb.DMatrix(data=X, label=y)
+    num_parallel_tree = 4
+    num_boost_round = 16
+    total_trees = num_parallel_tree * num_classes * num_boost_round
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply doc/python/model.rst:20]: Converted into a comment.

File: include/xgboost/c_api.h [hunk 5]
```diff
@@ -580,6 +580,21 @@ XGB_DLL int XGBoosterCreate(const DMatrixHandle dmats[],
...
 XGB_DLL int XGBoosterFree(BoosterHandle handle);
 
+/*!
+ * \brief Slice a model according to layers.
+ *
+ * \param handle Booster to be sliced.
+ * \param begin_layer start of the slice
+ * \param end_layer   end of the slice
+ * \param step        step size of the slice
+ * \param out Sliced booster.
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/c_api.h:590]: Added comments.

File: include/xgboost/gbm.h (hunk 1 above)
Comment (trivialfis) [reply include/xgboost/gbm.h:68]: Added comments.

File: tests/python/test_basic_models.py (hunk 2 above)
Comment (hcho3) [reply tests/python/test_basic_models.py:414]: Try `@pytest.mark.parameterize('booster',['gbtree', 'dart'])` (note the quotes around `booster`

Comment (hcho3) [reply tests/python/test_basic_models.py:414]: @trivialfis Also, `TestModels` should not be a subclass of `unittest.TestCase`. https://stackoverflow.com/a/35562401. Try making the class a subclass of `object`:
```python
class TestModels(object):
```

Comment (trivialfis) [reply tests/python/test_basic_models.py]: Thanks for the suggestion, done.

**Review by hcho3 — APPROVED** [review]

**Comment by hcho3** [comment]
The parametrization looks quite nice. The benefit is that, if a parametrized test fails for only one of the multiple parameters, the test result will clearly indicate which parameter led to failure. I'd like to submit a follow-up PR to introduce more test parametrization where it's appropriate. For example: https://github.com/dmlc/xgboost/blob/29745c6df2d6240579d9af22fd22e35db08acf02/tests/python/test_with_dask.py#L431-L444

**Pull request merged**